### PR TITLE
Multi-Arch Support

### PR DESCRIPTION
**Pull request by Scrumplex**
Implement multi-arch classifiers and OS-rules as per https://github.com/PolyMC/meta/pull/16

Supersedes #680 

# TODO
- [ ] Figure out a way to make "Download All" work when there is no cached Java metadata
- [ ] Make sure that we have all architecture mappings (See https://github.com/Scrumplex/PolyMC/blob/multi-arch-support/launcher/RuntimeContext.h#L13-L18)
- [ ] Make sure that we have all legacy architectures listed in https://github.com/Scrumplex/PolyMC/blob/multi-arch-support/launcher/RuntimeContext.h#L33

### CODE REVIEW

**Comment by DioEgizio** [comment]
You might wanna change pr name?

**Comment by Scrumplex** [comment]
I fucking hate GitHub. Sometimes it remembers the title when you reload the page, sometimes it doesn't :|

**Comment by r58Playz** [comment]
Might want to add `x86` to the list of legacy archs.
See https://github.com/r58Playz/meta-polymc/blob/master/org.lwjgl3/3.3.1.json#L66-L83

**Comment by DioEgizio** [comment]
that's not an issue because it's on 3.3.1, which uses the new split natives system

**Comment by Scrumplex** [comment]
The actual important part is how Java reports architecture. In my Windows VM it was reported as `x86` for latest Adoptium

**Comment by Scrumplex** [comment]
These should be all valid `os.arch` values for Windows: https://github.com/openjdk/jdk/blob/cb37282a12698ae66c27889db9251a5b278624b0/src/java.base/windows/native/libjava/java_props_md.c#L580-L588

Edit: Seems like older versions reported as `ia64`. Probably very unlikely today, but why not add that as well: https://github.com/openjdk/jdk/blob/9a9add8825a040565051a09010b29b099c2e7d49/jdk/src/windows/native/java/lang/java_props_md.c#L468-L476

File: launcher/RuntimeContext.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        system = currentSystem();
+    }
+
+    QString getClassifier() const {
+        return system + "-" + mappedJavaRealArchitecture();
+    }
+
+    // "Legacy" refers to the fact that Mojang assumed that these are the only two architectures
+    bool isLegacyArch() const {
+        QSet<QString> legacyArchitectures{"amd64", "x86_64", "ia64", "i386", "i686", "x86"};
```
Comment (Scrumplex) [inline launcher/RuntimeContext.h]: ia64 is not x86

**Comment by Scrumplex** [comment]
Rebased and resolved conflicts. This is probably ready to go. I will keep the TODO items unresolved, but maybe we can figure them out down the line

**Comment by chexo3** [comment]
Hey, this PR interests me. Could y'all review it sometime soon?

**Comment by TheCodex6824** [comment]
If you want more architectures for completeness: Java reports ppc64le for 64-bit little-endian PowerPC / Power ISA, while just ppc64 is **big endian** PowerPC / Power ISA. I know at least ppc64le works (I use that) with custom natives.

Also, [IA-64 is a different architecture](https://en.wikipedia.org/wiki/IA-64) and not x86 based, despite the naming. MS had a port of Windows to IA-64 for a while, but they eventually dropped it.

**Comment by chexo3** [comment]
Oh yeah, good notes. +1 on PPC64LE natives, I don’t personally own one but I know they’re becoming more available to people. For example Raptor Computing Systems makes some that folks like, and of course IBM sells them too.

**Review by timoreo22 — APPROVED** [review]

**Comment by chexo3** [comment]
OooOOoooOoOooO

You wanna merge this PR

oOOOoooOOoOOoo

You wanna review and approve it and then release a new package on the AUR

**Review by flowln — APPROVED** [review]
the code looks good to me!
seems like you forgor some of your beloved license headers though :p

**Comment by Scrumplex** [comment]
> seems like you forgor some of your beloved license headers though :p

impossible!

**Pull request merged**